include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -2147,55 +2317,71 @@ namespace internal
-                    const unsigned int fe_no_in,
+                    const unsigned int dof_no,
+                    const unsigned int first_selected_component,
                     const unsigned int quad_no_in,
                     const unsigned int fe_degree,
-                    const unsigned int n_q_points)
+                    const unsigned int n_q_points,
+                    const bool is_interior_face)
   :
   scratch_data_array (data_in.acquire_scratch_data()),
   quad_no            (quad_no_in),
-  n_fe_components    (data_in.get_dof_info(fe_no_in).n_components),
+  n_fe_components    (data_in.get_dof_info(dof_no).n_components),
   active_fe_index    (fe_degree != numbers::invalid_unsigned_int ?
-                      data_in.get_dof_info(fe_no_in).fe_index_from_degree(fe_degree)
+                      data_in.get_dof_info(dof_no).fe_index_from_degree(fe_degree)
                       :
                       0),
   active_quad_index  (fe_degree != numbers::invalid_unsigned_int ?
-                      (data_in.get_mapping_info().cell_data[quad_no_in].
+                      (is_face ?
+                       data_in.get_mapping_info().face_data[quad_no_in].
+                       quad_index_from_n_q_points(n_q_points)
+                       :
+                       data_in.get_mapping_info().cell_data[quad_no_in].
                        quad_index_from_n_q_points(n_q_points))
                       :
                       0),
+  n_quadrature_points(fe_degree != numbers::invalid_unsigned_int ? n_q_points :
+                      (is_face ?
+                       data_in.get_shape_info
+                       (dof_no, quad_no_in, active_fe_index, active_quad_index).n_q_points_face
+                       :
+                       data_in.get_shape_info
+                       (dof_no, quad_no_in, active_fe_index, active_quad_index).n_q_points)),
   matrix_info        (&data_in),
-  dof_info           (&data_in.get_dof_info(fe_no_in)),
-  mapping_data       (internal::MatrixFreeFunctions::MappingInfoCellsOrFaces<dim,Number,false>::get
-                      (data_in.get_mapping_info(), quad_no)),
+  dof_info           (&data_in.get_dof_info(dof_no)),
+  mapping_data       (internal::MatrixFreeFunctions::MappingInfoCellsOrFaces<dim,Number,is_face>::get(data_in.get_mapping_info(), quad_no)),
   data               (&data_in.get_shape_info
-                      (fe_no_in, quad_no_in, active_fe_index,
+                      (dof_no, quad_no_in, active_fe_index,
                        active_quad_index)),
   jacobian           (nullptr),
   J_value            (nullptr),
+  normal_vectors     (nullptr),
+  normal_x_jacobian  (nullptr),
   quadrature_weights (mapping_data->descriptor[active_quad_index].quadrature_weights.begin()),
   cell               (numbers::invalid_unsigned_int),
+  is_interior_face      (is_interior_face),
+  face_vector_access_index (is_face ? (is_interior_face ? 0 : 1) : 2),
```

Comment:
maybe make enumerator if it only takes 3 values?

Reply:
(same below)

Reply:
Let's postpone this to a later PR where the variable is actually used. I have deleted it for now.

Reply:
fair enough.